### DISCUSSION TITLE
Copilot Chat and GPT-4o

### DISCUSSION THREAD

**Discussion by pidgeon777** [community/community#123670] [Copilot Conversations]
### Select Topic Area

General

### Body

With the recent public unveiling of the GPT-4o model by OpenAI, there has been considerable interest in its potential applications. One area of particular interest is its integration with GitHub Copilot, especially in the context of enhancing the chat functionality.

Given the capabilities of GPT-4o, its adoption could significantly improve the quality of code generation, provide more accurate responses, and offer a more interactive and engaging user experience. 

However, as of now, there is no official information available regarding the timeline or the plan for the integration of GPT-4o into GitHub Copilot. 

If anyone has any insights or updates about this, it would be greatly appreciated if you could share them. Thank you!

**Comment by maikmetzger** [comment]
I hope we'll see it soon. Even as an experimental feature, where you can select between GPT-4 Turbo and GPT-4o for the time being would be cool. I know downward-compatibility for new features is a pain to maintain, therefore GPT-4 Turbo should be deprecated once ensured GPT-4o is working as intended without any drawbacks. That's what Nightly release is for, isn't it?
I hope it will be a lot faster now than waiting for GPT-4 in GitHub Copilot in the past.

**Reply by pidgeon777** [reply]
I hope so. Given that GPT-4o already seems to work very well, it will hopefull take a short time to also update Copilot with the same model. All of this responsiveness introduced with GPT-4o would be a tremendous plus while coding.

**Reply by H4lis** [reply]
harapannya semoga free

**Comment by cannap** [comment]
Yes, I can't wait. Maybe they'll add a switch to toggle between GPT-4 and GPT-4o. But I know they already knew that GPT-4o was coming. :D

**Reply by H4lis** [reply]
betul

**Comment by pidgeon777** [comment]
Honestly I've observed a significant improvement when transitioning from model 4 to 4o. This is particularly noticeable when using Lua for programming plugins for my Neovim editor, and when using Python for other tasks. 

With the 4th model, the success rate of solutions working immediately was around 50-60%. However, with the latest model, I believe this has increased to 80%. Additionally, the speed of generating solutions has noticeably improved. 

Furthermore, the latest model appears to employ more advanced techniques, enhancing the overall quality and efficiency of the solutions generated.

**Reply by Seikilos** [reply]
Can you elaborate on this? How is your workflow for coding, where 4o does help you? Do you paste code into the chat interface, are you using the API?

Because for me the gap of usability between IDE integration and using a website chat interface is huge.

**Reply by pidgeon777** [reply]
My workflow involves copying and pasting the code I want to update into the ChatGPT prompt.

I then specify the desired action, such as refactoring, adding comments, improving a loop, etc. I've observed that when the code becomes too large, it may lose some information or even parts of the previous code. Therefore, it's always important to check for newly added modifications.

However, when the code is not overly complex, the solutions provided often work right out of the box.

I tested it mainly with Lua and Python.

**Comment by aurexav** [comment]
Yes, I've disabled the copilot for now.
GPT-4o is incredibly fast and smart! I hope to use it directly in VSCode soon!

**Comment by mohammad-malik** [comment]
Would love an option to choose between models! GPT-4o is extremely smart for fixing and adjusting code, but it does hallucinate often, particularly with information on other frameworks that are not coding, such as Docker.

**Comment by denizsincar29** [comment]
I emagine having a voice assistent GPT-4O in VSCode. Asking hey code, what would you suggest here? she answers: i suggest to add a check for int input here.. As I move the cursor, it will guide me on everything I do, even inside VSCode, palates, commits and everything I do also outside the workspace.

**Comment by mpittas** [comment]
I hope they release it soon enough. In my opinion it shouldn't take more than a month or two as developers like me will move to other coding assistants that already provide the GPT-4o model.

**Comment by andrew-tiosa** [comment]
I just realized that the knowledge is cured off 2021. im mean I can believe that.

**Reply by aswer1337** [reply]
lol

**Comment by adnanbaqi** [comment]
OpenAI and GitHub Announcements: Keep an eye on official announcements from OpenAI and GitHub. These are typically the primary sources for news regarding updates to products like GitHub Copilot.
Tech Blogs and Media: Technology news websites and blogs often cover significant updates about popular tools and services. Following these can provide early insights into new integrations and features.
Developer Forums and Communities: Engaging with communities on platforms like Reddit, Stack Overflow, or even GitHub's own community forums can be useful. Members often share updates, beta releases, and their own experiences with new technologies.
Subscribe to Newsletters: Many tech companies and media outlets offer newsletters that include updates on software tools and AI advancements. Subscribing to those related to OpenAI or software development can keep you informed.
Attend Tech Conferences and Webinars: These events often include announcements about new technologies and partnerships. They can also be a good opportunity to ask representatives directly about upcoming features and integrations.

**Comment by jamesfricker** [comment]
Copilot is now quite a few generations behind the latest models. Even having just the chat mode as the latest model would be a massive upgrade.

**Reply by bazzilic** [reply]
Isn't Copilot currently based on GPT-4? And AFAIK GPT-4o is more or less same intelligence as GPT-4, its main feature is it's faster.

**Reply by Al4ric** [reply]
Logic is much improved in 4o. I have checked myself. Recently I had issue in Angular and GitHub Copilot was helping me to fix this. After 10 or more prompts it still was not working. I pasted issue into Chat GPT 4o and got correct answer on first try.

**Reply by bazzilic** [reply]
ChatGPT on OpenAI website is way smarter than Copilot Chat, that's true, always has been. I'm quite sure that even the 3.5 model on the website is smarter that Copilot Chat. Maybe Copilot Chat uses a dedicated faster and stupider model, maybe it's because of the Copilot prompt.

**Comment by Seikilos** [comment]
Not having copilot here atm, so have to ask: can somebody try to prompt the code gen feature to leak which model it uses?

Like `// generate a console output that tells me your chat gpt model, that you are using`?

**Reply by bazzilic** [reply]
idk if that's true though

<img width="854" alt="image" src="https://github.com/community/community/assets/1805832/ff58da6d-fe21-4d5a-b449-ddfa788062fe">

**Reply by Al4ric** [reply]
This is from chat: console.log("The chat GPT model version in use is GPT-4");
This is from inline chat: print("I am using the GPT-3 model for chat.")

**Reply by Seikilos** [reply]
Tricky and not reliable, but we know, that none of the AIs claim to be 4o.

**Reply by cannap** [reply]
> Tricky and not reliable, but we know, that none of the AIs claim to be 4o.

is asked 10000 times sometimes it tells 4 sometimes it tells 3.5

**Reply by penglei0** [reply]
It uses gpt-3.5 or gpt-4 turbo. You find the info in the log of copilot .

**Reply by aurexav** [reply]
Bro, what's your font? It looks not bad! @bazzilic

**Reply by bazzilic** [reply]
@AurevoirXavier Cascadia Code PL (https://github.com/microsoft/cascadia-code)

**Comment by denisdasilvarocha** [comment]
2 Weeks and still no GPT-4o? What the hell is wrong with this product, jesus.. thank god i cancled my sub 7,8 Months ago Copilot is still so much behind insane.

**Comment by 1saptarshi** [comment]
#1 Enhanced Understanding and Contextualization
#2  Greater Accuracy and Reliability
#3 Advanced Reasoning Capabilities
#4 Multimodal Capabilities
#5 Improved Fine-Tuning and Customization
#6 Reduced Bias and Enhanced Safety
#7 Scalability and Efficiency
#8 Enhanced User Experience

**Comment by vfonic** [comment]
I'd love to be able to choose a model. Claude Sonnet 3.5 seems to be the best model for generating code atm.

**Reply by 001123** [reply]
Switch to [cody ai](https://sourcegraph.com/cody), it has more choice

**Comment by pidgeon777** [comment]
Surprisingly, there is still no news about the integration of GPT-4o into Copilot Chat. I would have expected an announcement by now, which seems to be delayed.

**Comment by asare4u** [comment]
There's no official word on GPT-4o in Copilot yet, but there's definite interest!  People are excited about the potential for better code generation and a more interactive experience.  Since GPT-4o is open source, it aligns with GitHub's approach, so future Copilot updates might integrate advancements from similar models.

**Reply by pidgeon777** [reply]
> There's no official word on GPT-4o in Copilot yet, but there's definite interest! People are excited about the potential for better code generation and a more interactive experience. Since GPT-4o is open source, it aligns with GitHub's approach, so future Copilot updates might integrate advancements from similar models.

I was under the impression that the GPT-4o model was not open source. Could you please provide me with some references or resources where I can learn more about this? Additionally, could you clarify what you mean by "open source" in this context? I would like to understand your perspective better.

**Reply by vadim-su** [reply]
I even double-checked this information. No, GPT-4o is not open source.

**Reply by asare4u** [reply]
My apologies for the earlier confusion! GPT-4o isn't currently open source. There might be misunderstandings online about an open version.

To clarify: Open source means the code is freely available for everyone to see and work with. While there's no official open-source GPT-4o, there are discussions about open-source alternatives that offer similar functionalities.

**Reply by vfonic** [reply]
GPT-4o is not _currently_ open source nor will it ever be.

**Reply by pidgeon777** [reply]
> My apologies for the earlier confusion! GPT-4o isn't currently open source. There might be misunderstandings online about an open version.
> 
> To clarify: Open source means the code is freely available for everyone to see and work with. While there's no official open-source GPT-4o, there are discussions about open-source alternatives that offer similar functionalities.

I find it unlikely that developers here on GitHub do not know the difference between "open source" and "non-open source". What I'm wondering is: are you perhaps a bot? Your answers sound "strange". Maybe it's just my impression.
I recognize something of ChatGPT in your responses.

**Reply by asare4u** [reply]
I'm not a bot @pidgeon777. I am a lad who wants to know more about tech. That's all.

**Reply by retrievaaaa** [reply]
> My apologies for the earlier confusion! GPT-4o isn't currently open source. There might be misunderstandings online about an open version.
> 
> To clarify: Open source means the code is freely available for everyone to see and work with. While there's no official open-source GPT-4o, there are discussions about open-source alternatives that offer similar functionalities.

ChatGPT lying about its own models being open source is kinda hilarious

**Reply by retrievaaaa** [reply]
> > My apologies for the earlier confusion! GPT-4o isn't currently open source. There might be misunderstandings online about an open version.
> > To clarify: Open source means the code is freely available for everyone to see and work with. While there's no official open-source GPT-4o, there are discussions about open-source alternatives that offer similar functionalities.
> 
> I find it unlikely that developers here on GitHub do not know the difference between "open source" and "non-open source". What I'm wondering is: are you perhaps a bot? Your answers sound "strange". Maybe it's just my impression. I recognize something of ChatGPT in your responses.

No doubts, he definitely wrote that answer using AI.

**Comment by jakearlen879** [comment]
GitHub Copilot Enterprise is on GPT-4o: https://github.blog/changelog/2024-07-05-github-copilot-enterprise-on-gpt-4o

**Reply by tgpsantos** [reply]
has anyone tried it? what's the feedback? it's quite an increase in terms of pricing, so wondering what the quality and speed looks like compared to the regular copilot

**Comment by pidgeon777** [comment]
Recently, on July 7, a user named `JiaSing Lin` shared a post in this discussion. However, it seems that the post has been removed.

![Image of the removed post](https://github.com/community/community/assets/13739892/0dd6a648-fd99-4740-ac3f-12821b3bab1e)

One wonders what happened to `JiaSing Lin`'s post.

Currently, **GitHub Copilot** offers three subscription plans:

1. **Individual**: $10 per month
2. **Business**: $19 per month
3. **Enterprise**: $39 per month

From what it seems, `JiaSing Lin` has a **Personal** subscription but seems to have access to the `GPT-4o` model. This raises the question: will it be possible to use `GPT-4o` with the **Personal** plan in the future?

If not, it would be a great disappointment. There are alternatives to the `GPT-4o` model, such as `claude.ai`, which offer superior performance at much more affordable costs compared to the **Business** plan.

We hope that `JiaSing Lin`'s access to the `GPT-4o` model is not an isolated case and that soon even users with a **Personal** plan can benefit from this feature.

**Reply by anthonyy41** [reply]
Hi pidgeon777,

Thank you for your concern regarding my previous message. To ensure that the knowledge cutoff date for the same personal account is consistent across all devices, I conducted a cross-test by logging into the same personal account from another computer and inquired about the knowledge cutoff date for GPT-4o. Unfortunately, I did not receive an answer specific to the GPT-4o cutoff date.

To avoid any misunderstanding or confusion about the personal version supporting GPT-4o, I decided to remove my previous message and screenshots.

Additionally, I completely agree with your viewpoint and sincerely hope that the GitHub Team updates the model for the personal plan to GPT-4o soon.

**Reply by pidgeon777** [reply]
Hello @anthonyy41,

Thank you for your comprehensive response.

We are hopeful that the **Personal** plan of **GitHub** will receive the update soon. Currently, there are alternatives to **GitHub**, and while I have no immediate plans to migrate to other platforms, I may be compelled to do so if necessary.

I trust that the **GitHub** team will heed our requests. After all, the "common" users of **GitHub**, who contribute significantly to enriching the platform's codebase, should be entitled to the same benefits as those reserved for **Enterprise** plan users. Despite being more expensive, the latter does not necessarily offer features that would be utilized by the majority of users.

Additionally, I'm curious to know if you've observed any performance improvements when using the `GPT-4o` model as compared to `GPT-4`. If so, could you share the specific improvements you've noticed? For instance, increased speed, more accurate and comprehensive responses, etc.

Lastly, if you prefer, I can remove the screenshot I took of your message. If not, I will leave it accessible for anyone else who might be interested.

**Reply by anthonyy41** [reply]
hi @pidgeon777 

Of course, no problem. You can keep the relevant information from my previous messages!

From my experience, I’ve noticed several advantages when using GitHub Copilot Chat paired with GPT-4o:

1. Updated Knowledge: GPT-4o has a knowledge cutoff date of October 2023, whereas GPT-4 is September 2021. This allows it to provide more current frameworks and technical information, resulting in more practical suggestions.

2. Improved Speed and Quality: The new model offers enhanced processing speed and higher quality responses for complex problems, minimizing the need for multiple interactions to achieve the desired code examples.

3. Better Contextual Understanding: With a higher token limit, GPT-4o can more comprehensively understand an entire project's structure, offering more relevant and structured code suggestions.

I hope this feedback is helpful to you! :)

**Reply by pidgeon777** [reply]
Very useful 👍, thank you @anthonyy41. Hopefully this will shortly come also for the **Personal** plan.

**Comment by ericktreichel** [comment]
I believe more transparency from Microsoft/Github about the models used to power Github Copilot services at any given time would be really good for us.

I mean, why make it harder for us to get a better idea about what we should expect from the outputs that they provide to our code?
Coding is a serious business. It would be nice to know if the suggestion that Copilot is giving me was generated from a less powerfull model like the GPT3.5 Turbo or a more powerfull GPT4-o.

In a recent post (https://azure.microsoft.com/en-us/blog/openais-fastest-model-gpt-4o-mini-is-now-available-on-azure-ai/) about the release of GPT4-o mini Microsoft says something about noticing an "incredible speed while testing GPT-4o mini on GitHub Copilot", but they don't say if it will be released soon on Github Copilot, if it is already being used, if it will never be used, nothing!

Why not more transparency?

**Comment by retrievaaaa** [comment]
I feel like Copilot has become significantly better out of nowhere, maybe they switched to 4o-mini now?

**Reply by pidgeon777** [reply]
Hello, what do you mean with better? I have not a chance to test it right now, do you mean:

- Faster
- No more laziness (stopping itself before completing the answer)
- Increased input / output tokens (accepts more input words, lengthier answers etc.)
- Increased knowledge cutoff (e.g. it knows about events till mid 2024)
- Less hallucinations
- Output code with less errors

I am really curious to know which of the above points you're referring to.

**Reply by retrievaaaa** [reply]
For me, it seems like it uses my code as context way better now, which might relate to the larger context window of 4o mini; also it seems that it auto completes mid-line better, like I often had it complete

`if(blablabla || °) { `(° being the cursor pos)
to
`if(blablabla || blabla) {) {`
but now it correctly completes it to
`if(blablabla || blabla) {`

but that might just be improved handling of the output

It **_feel like_** it is faster, but can't actually confirm that without actually comparing them.

**Comment by parvic1** [comment]
They should at least provide a road map for the model implementation on Copilot.

**Reply by pidgeon777** [reply]
I agree.

**Comment by jamesfricker** [comment]
https://github.blog/changelog/2024-07-31-github-copilot-chat-and-pull-request-summaries-are-now-powered-by-gpt-4o/


Looks like it's using gpt4o now - although my chat windows still says it's using Turbo even after updating to the latest version 🤔

EDIT: updating vscode fixed the issue

**Reply by HoverCatz** [reply]
There's no way this is using GPT-4o. It's such a letdown. Even with all the hype, it still sucks. I bought it because of the news, but it's just as bad as when I uninstalled it months ago. The free GPT-4o version on OpenAI’s website is way better.

I'm using the latest plugin version on JetBrains, and it keeps giving me duplicate code. It doesn't listen to my requests, like not using my existing code and writing something totally new. It keeps using my old code anyway. The responses are super basic, with hardly any explanation, and it feels like I'm still stuck with GPT-3.

In my case, it would be false advertising. I actually feel scammed.

Edit:
I just asked it to add a default font (for the MonoGame library), and instead of just using an existing one, it started creating one from scratch:
[Imgur](https://imgur.com/a/YjeucHi)
It spammed 186 lines of the same empty Vector3 until it timed out. Very good GPT-3 behaviour right there.

**Reply by jamesfricker** [reply]
Updating my vscode has fixed my issue. Chat is using 4o model now

**Reply by pidgeon777** [reply]
Copilot Chat has been upgraded to 4o model.

What about GitHub Copilot? Which model is it using? I'm referring to the Copilot autocomplete, not chat.

**Reply by eminSerin** [reply]
Of course, I am not sure of the details, but Copilot (not Chat) still uses copilot-codex API.

**Reply by avxkim** [reply]
> Copilot Chat has been upgraded to 4o model.
> 
> What about GitHub Copilot? Which model is it using? I'm referring to the Copilot autocomplete, not chat.

Have you compared native 4o with the Copilot's chat version? how was results?

**Reply by HoverCatz** [reply]
I have no idea how you guys are using the new gpt4o version. Mine is actually retarded. It keeps repeating shit and is just fucking annoying to use. No way this is gpt4o. Has to still be gpt3.

**Reply by savanesoff** [reply]
Interesting, I'm using latest vscode and its still a Turbo for me. is this limitted to regions?

**Reply by savanesoff** [reply]
> Interesting, I'm using latest vscode and its still a Turbo for me. is this limitted to regions?

Nevermind, had to rrestart VSCode again. :smile:

**Comment by hidaris** [comment]
Yes, Chat is now at 4o, but I'm more concerned about when the completion update will roll out for 4o mini

**Reply by ericktreichel** [reply]
Did they say something about changing the model for completion? 
I guess they will update the Codex itself to rely on 4o mini, in this case. I hope they do...

**Reply by parvic1** [reply]
how do you know? I keep asking Copilot and keep getting GPT-4.

**Reply by cannap** [reply]
> how do you know? I keep asking Copilot and keep getting GPT-4.

https://code.visualstudio.com/updates/v1_92#_github-copilot

**Reply by eminSerin** [reply]
> how do you know? I keep asking Copilot and keep getting GPT-4.

Very, because it basically tells me it's using GPT 4-o.

**Reply by vadim-su** [reply]
does completion still work on gpt-3?

**Reply by pidgeon777** [reply]
I'm also interested in knowing that.

**Reply by pidgeon777** [reply]
> does completion still work on gpt-3?

Maybe one day GPT-4o mini version will be used for Copilot completion? Perhaps because it is more suitable as a model, but I imagine some tests will need to be conducted.

**Reply by amnweb** [reply]
Definitely chat is not at GPT-4o, too many wrong answers on same prompt between copilot chat and GPT-4o, I'm 100% sure that chat is still on GPT4

**Comment by HoverCatz** [comment]
I said it twice already and I will say it again.
There is no way this is gpt-4o. It's literal **SHIT**. Github Copilot Chat doesn't even come close to openai's gpt-4o.

**Reply by clay-rip** [reply]
most likely 4o mini

**Reply by pidgeon777** [reply]
> most likely 4o mini

How could one obtain proof of this?

**Reply by AleksandrSemykin** [reply]
I asked Copilot Chat which model it is using and it says "I am using the GPT-4 Turbo model." (VSCode GitHub Copilot extention). 
Officially GPT-4o was available weeks ago, but it is not working in my case.

**Reply by markox92** [reply]
But they promote GPT-4o in VSCode extension, So you will pay for GPT-4o and you will get old one, hmm this is ripe for a lawsuit

**Reply by AleksandrSemykin** [reply]
The GitHub Support confirmed that GPT-4o is the current model and Copilot Chat is not aware of its actual model.

**Reply by clay-rip** [reply]
It's technically GPT-4o, but copilot is way too stupid to be full 4o. It has to be 4o mini.

**Reply by pidgeon777** [reply]
I'm not sure why it should technically be `GPT-4o`, but being in effect the `mini` version. That would mean they're lying? But it seems the GitHub Support confirmed the `GPT-4o` version instead.
Still there seems to be differences between the **ChatGPT Web** and **Copilot Chat** model instances (the former seems to be better than the latter, even though they use both models apparently).

**Reply by clay-rip** [reply]
Microsoft lying wouldn't be a first

**Reply by ForareVictor** [reply]
Anyone else that get the same results as me? 
They ruined it with the last couple of updates, it is 100% worse than when I started using it, it fail to put more than a handful of methods into context when giving answers and it is no longer possible to have a conversation with it about a solution, it's like talking to a wall, it's just very robotic and shits out random code, ignoring most of what we tried before or discussed, ( all this worked some months ago)

**Reply by pidgeon777** [reply]
It's hard to say in my case, it seems to work good, but someway with a lower performance compared to the web version of ChatGPT with the GPT-4o model.

**Comment by akudiaku** [comment]
https://potential-xylophone-q7qxwp7jx97wf9544.github.dev/MyAnPoinDatabase projects ```
@
```

**Comment by akudiaku** [comment]
https://potential-xylophone-q7qxwp7jx97wf9544.github.dev/MyAnPoinDatabase projects ```
@
```

**Comment by pidgeon777** [comment]
> https://potential-xylophone-q7qxwp7jx97wf9544.github.dev/MyAnPoinDatabase projects ```
> @
> ```

What's that?

**Reply by parvic1** [reply]
I have reported this content. Seems fishy.

**Reply by akudiaku** [reply]
🤔??

**Comment by akudiaku** [comment]
https://potential-xylophone-q7qxwp7jx97wf9544.github.dev/MyAnPoinDatabase projects ```
@
```@akudiaku

**Comment by akudiaku** [comment]
https://potential-xylophone-q7qxwp7jx97wf9544.github.dev/MyAnPoinDatabase projects ```
@
```@akudiaku 
    By administrator

**Comment by Kevin-Pro-Codes** [comment]
hows is it trully? which model?
